Sources/MapboxNavigation/NavigationMapView+BuildingHighlighting.swift

```diff
@@ -11,17 +11,93 @@ extension NavigationMapView {
      
      - parameter coordinates: Coordinates which represent building locations.
      - parameter extrudesBuildings: Switch which allows to highlight buildings in either 2D or 3D. Defaults to true.
-     
-     - returns: Bool indicating if number of buildings found equals number of coordinates supplied.
+     - parameter completion: A closure that operates after the buildings are highted.
+
      */
-    @discardableResult public func highlightBuildings(at coordinates: [CLLocationCoordinate2D], in3D extrudesBuildings: Bool = true) -> Bool {
-        return false
+    public func highlightBuildings(at coordinates: [CLLocationCoordinate2D],
+                                   in3D extrudesBuildings: Bool = true,
+                                   completion: () -> Void) {
+        let group = DispatchGroup()
+        guard let identifiers = try? mapView.__map.getStyleLayers().compactMap({$0.id}).filter({ $0.contains("building") }) else { return }
+
+        let _ = coordinates.map({
+            group.enter()
+            let screenCoordinate = mapView.point(for: $0, in: self)
+            mapView.visibleFeatures(at: screenCoordinate,
+                                    styleLayers: Set(identifiers),
+                                    completion: { [weak self] result in
+                                        guard let validSelf = self else { return }
+                                        if case .success(let features) = result {
+                                            if let identifier = features.first?.featureIdentifierToDouble() {
+                                                self?.foundBuildingIds.insert(identifier)
+                                            }
+                                            group.leave()
+                                        }})
+        })
+
+        group.notify(queue: DispatchQueue.main) {
+            self.addBuildingsLayer(with: self.foundBuildingIds, in3D: extrudesBuildings)
+        }
     }
     
     /**
      Removes the highlight from all buildings highlighted by `highlightBuildings(at:in3D:)`.
      */
     public func unhighlightBuildings() {
-        
+        guard let _ = try? mapView.style.getLayer(with: IdentifierString.buildingExtrusionLayer, type: FillExtrusionLayer.self).get() else { return }
+        foundBuildingIds = Set<Double>()
```

Comment:
Why do you keep exposing `foundBuildingIds` as a member variable? It can be stored locally as client doesn't need it. Instead completion handler should have boolean parameter to check whether all identifiers were found.

Reply:
Thanks for the suggestion, I'll change it to local.